libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -219,109 +265,200 @@ BeaconSetupService::State BeaconSetupService::OnWaitForDirectConnections()
+  FETCH_LOG_DEBUG(LOGGING_NAME, "Relevant connections: ", connections.size());
+
+  beacon_dkg_connections_gauge_->set(connections.size());
+
+  // If we get over threshold connections, send a message to all peers each time
+  // we increase over this threshold (note we won't advance if we ourselves don't get over)
+  if (connections.size() > connections_.size() && connections.size() >= require_connections)
+  {
+    fetch::serializers::MsgPackSerializer serializer;
+    serializer << connections_;
+    pre_dkg_rbc_.Broadcast(serializer.data());
+
+    FETCH_LOG_INFO(LOGGING_NAME, "Minimum peer threshold requirement met for DKG");
+
+    connections_ = ConvertToSet(connections);
+  }
+
+  // Whether to proceed (if threshold peers have also met this condition)
+  const bool is_ok = (ready_connections_.size() >= require_connections &&
+                      connections_.size() >= require_connections);
+
+  if (!condition_to_proceed_ && is_ok)
+  {
+    condition_to_proceed_ = true;
+    FETCH_LOG_INFO(LOGGING_NAME, "State: ", ToString(state_machine_->state()),
+                   " Ready. Seconds to spare: ", state_deadline_ - GetTime(), " of ",
+                   seconds_for_state_);
+  }
+
+  if (timer_to_proceed_.HasExpired())
+  {
+    if (!condition_to_proceed_)
+    {
+      FETCH_LOG_WARN(LOGGING_NAME, "Failed to guarantee peers were ready for DKG!");
+      SetTimeToProceed(State::RESET);
+      return State::RESET;
     }
-    FETCH_LOG_INFO(LOGGING_NAME, "All peers connected. Proceeding.");
 
     BroadcastShares();
-    return State::WAIT_FOR_SHARE;
+    SetTimeToProceed(State::WAIT_FOR_SHARES);
+    return State::WAIT_FOR_SHARES;
+  }
+  else
+  {
+    if (!condition_to_proceed_)
+    {
+      FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to be ready before starting DKG. Ready: ",
+                     connections.size(), " expect: ", require_connections);
+    }
+
+    state_machine_->Delay(std::chrono::milliseconds(500));
   }
 
-  state_machine_->Delay(std::chrono::milliseconds(100));
-  FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to be ready before starting DKG. Ready: ",
-                 ready_connections_.size(), " expect: ", beacon_->aeon.members.size() - 1);
   return State::WAIT_FOR_READY_CONNECTIONS;
 }
 
+/**
+ * The node has broadcast it's own shares at this point, now wait
+ * to receive shares from everyone else
+ *
+ */
 BeaconSetupService::State BeaconSetupService::OnWaitForShares()
 {
   std::lock_guard<std::mutex> lock(mutex_);
-  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_SHARE));
+  beacon_dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_SHARES));
+
+  const bool coeff_failed  = (coefficients_received_.size() < beacon_->manager.polynomial_degree());
+  const bool shares_failed = (shares_received_.size() < beacon_->manager.polynomial_degree());
+  const bool is_ok         = !coeff_failed && !shares_failed;
```

Comment:
Ok, it will become:

```
  auto intersection = (coefficients_received_ & shares_received_);
  const bool is_ok         = intersection.size() >= beacon_->manager.polynomial_degree();
```